### PR TITLE
GEOMESA-2544 Composite XML converter

### PR DESCRIPTION
**Pull request by elahrvivaz**
* Adds 'xpath' converter function for use in predicates
* Literal expressions are now pre-evaluated during expression parsing

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

**Comment by elahrvivaz** [comment]
note: stacks on top of #2116

**Comment by jnh5y** [comment]
Is this ready to be rebased?

File: geomesa-convert/geomesa-convert-xml/src/main/scala/org/locationtech/geomesa/convert/xml/XmlConverter.scala [hunk 1]
```diff
@@ -39,69 +39,23 @@ import scala.util.control.NonFatal
...
-    }
-    xp
-  }
+  private val xpath = XmlConverter.createXPath(config.xpathFactory, config.xmlNamespaces)
 
   private val rootPath = config.featurePath.map(xpath.compile)
 
   fields.foreach(_.compile(xpath))
 
   // TODO GEOMESA-1039 more efficient InputStream processing for multi mode
```
Comment (jnh5y) [inline geomesa-convert/geomesa-convert-xml/src/main/scala/org/locationtech/geomesa/convert/xml/XmlConverter.scala]: Does this comment need to move?

File: geomesa-convert/geomesa-convert-xml/src/test/scala/org/locationtech/geomesa/convert/xml/XmlCompositeConverterTest.scala [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        |  ]
+        |}
+      """.stripMargin)
+  )
+
+  "XmlCompositeConverter" should {
+
+    "select a parser based on a predicate" >> {
+      val xmlStr =
+        """<doc><type>a</type><id>1</id><number>123</number><color>red</color><physical><weight>127.5</weight><height>5'11"</height></physical><lat>0</lat><lon>0</lon></doc>
```
Comment (jnh5y) [inline geomesa-convert/geomesa-convert-xml/src/test/scala/org/locationtech/geomesa/convert/xml/XmlCompositeConverterTest.scala]: Does XML composite converters work by applying different converters to each file?  Or can one process sub-documents differently?

File: geomesa-convert/geomesa-convert-xml/src/main/scala/org/locationtech/geomesa/convert/xml/XmlConverter.scala (hunk 1 above)
Comment (elahrvivaz) [reply geomesa-convert/geomesa-convert-xml/src/main/scala/org/locationtech/geomesa/convert/xml/XmlConverter.scala]: ah, yeah it should probably go in the 'iterator' method, but i guess it's still relevant here where it's invoked...

File: geomesa-convert/geomesa-convert-xml/src/test/scala/org/locationtech/geomesa/convert/xml/XmlCompositeConverterTest.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-convert/geomesa-convert-xml/src/test/scala/org/locationtech/geomesa/convert/xml/XmlCompositeConverterTest.scala]: you can only apply a single converter to each xml doc, so you couldn't use a feature-path and handle each sub-document separately, if that's what you're asking.

Comment (elahrvivaz) [reply geomesa-convert/geomesa-convert-xml/src/test/scala/org/locationtech/geomesa/convert/xml/XmlCompositeConverterTest.scala]: if that's a use case, we can open a ticket to support it.

**Pull request merged**